functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+````clojure
+(defmacro  _transact [db op & txs]
+   (when txs
+     (loop [[frst-tx# & rst-tx#] txs  res#  [op db `transact-on-db]  accum-txs# []]
+       (if frst-tx#
+           (recur rst-tx# res#  (conj  accum-txs#  (vec frst-tx#)))
+           (list* (conj res#  accum-txs#))))))
+````
+* *swap!* invokes *transact-on-db* within a transaction (with the previously prepared arguments)
+* *transact-on-db* creates the new state of the database and returns it
+
+At this point we can see that with few minor tweaks, we can also provide a way to ask "what-if" questions. It can be done by replacing *swap!* with a function that would not impose any change to the system. This scenario is implemented with the "what-if" call chain:
+
+what-if → _transact →   _what-if → transact-on-db
+
+* The user calls *what-if* with the database value and the operations to perform. It then relays these inputs to *_transact*, adding to them a function that mimics *swap!*'s APIs, without its effect (callled *_what-if*).  
+````clojure
+(defmacro what-if [db & ops]  `(_transact ~db _what-if  ~@ops))
+````
+* *_transact* prepares the call to *_what-if*. It does so by creating a list that begins with *_what-if*, followed by the database, then the *transact-on-db* symbol and the batch of operations.
+* *_what-if* invokes *transact-on-db*, just like *swap!* does in the transaction scenario, but does not inflict any change on the system.
+
+````clojure
+(defn- _what-if [db f txs]  (f db txs))
+````
+ 
+Note that we are not using functions, but macros. The reason for using macros here is that arguments to macros do not get evaluated as the call happens, this allows us to offer a cleaner API design where the user provides the operations structured in the same way that any function call is structured in Clojure. 
+
+The above described process can be seen in the following examples:
+
+Transaction: 
+
+User call: 
+````clojure
+(transact db-conn  (add-entity e1) (update-entity e2 atr2 val2 :db/add))  
+````
+Changes into: 
+````clojure
+(_transact db-conn swap! (add-entity e1) (update-entity e2 atr2 val2 :db/add))
+````
+Becomes: 
+````clojure
+(swap! db-conn transact-on-db [[add-entity e1][update-entity e2 atr2 val2 :db/add]])
+````
+What-if: 
+
+User call: 
+````clojure
+(what-if my-db (add-entity e3) (remove-entity e4))
+````
+Changes into: 
+````clojure
+(_transact my-db _what-if (add-entity e3) (remove-entity e4))
+````
+Changes into: 
+````clojure
+(_what-if my-db transact-on-db [[add-entity e3] [remove-entity e4]])
+````
+Becomes eventually: 
+````clojure
+(transact-on-db my-db  [[add-entity e3] [remove-entity e4]])
+````
+
+## Insight extraction as libraries
+
+At this point, we have the core functionality of the database in place, and it is time to add to it its raison d'être - insights extraction. The architecture approach we used here is to allow adding these capabilities as libraries, as different usages of the database would need different such mechanisms. 

```

Comment:
It's a name for all kinds of querying capabilities, as querying is how to extract insight from a database (as oppose to extract data, which is what storage systems are for)